### PR TITLE
Match orders with a new user

### PR DESCRIPTION
**Pull request by d-wysocki**
closes [#5398](https://github.com/mirumee/saleor/issues/5398)

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations

# Pull Request Checklist

* [X] Privileged queries and mutations are guarded by proper permission checks
* [X] Database queries are optimized and the number of queries is constant
* [X] Database migration files are up to date
* [X] The changes are tested
* [X] GraphQL schema and type definitions are up to date
* [X] Changes are mentioned in the changelog

### CODE REVIEW

File: tests/api/test_account.py [hunk 1]
```diff
@@ -633,18 +636,21 @@ def test_customer_register(send_account_confirmation_email_mock, api_client):
...
 def test_customer_register_disabled_email_confirmation(
-    send_account_confirmation_email_mock, api_client
+    match_orders_with_new_user_mock, send_account_confirmation_email_mock, api_client
 ):
     variables = {"email": "customer@example.com", "password": "Password"}
     response = api_client.post_graphql(ACCOUNT_REGISTER_MUTATION, variables)
     errors = response.json()["data"]["accountRegister"]["errors"]
     assert errors == []
     assert send_account_confirmation_email_mock.delay.call_count == 0
+    assert match_orders_with_new_user_mock.call_count == 1
```
Comment (fowczarek) [inline tests/api/test_account.py:653]: We should use `assert_called_once_with` instead of call_count.

File: tests/test_order_utils.py [hunk 2]
```diff
@@ -44,3 +44,20 @@ def test_change_quantity_generates_proper_event(
...
+    address = customer_user.default_billing_address.get_copy()
+    Order.objects.create(
+        billing_address=address, user=None, user_email=customer_email,
+    )
+
+    assert get_order_user_by_customer_email(customer_email) is None
+
+    match_orders_with_new_user(customer_user)
+
+    assert get_order_user_by_customer_email(customer_email) == customer_user
```
Comment (fowczarek) [inline tests/test_order_utils.py:63]:
```suggestion
    order = Order.objects.create(
        billing_address=address, user=None, user_email=customer_email,
    )

    match_orders_with_new_user(customer_user)
    order.refresh_from_db()
    assert order.user == customer_user
```

I think this test is more readable.

File: tests/api/test_account.py [hunk 3]
```diff
@@ -633,18 +636,21 @@ def test_customer_register(send_account_confirmation_email_mock, api_client):
     customer_creation_event = account_events.CustomerEvent.objects.get()
     assert customer_creation_event.type == account_events.CustomerEvents.ACCOUNT_CREATED
     assert customer_creation_event.user == new_user
+    assert match_orders_with_new_user_mock.call_count == 0
```
Comment (fowczarek) [inline tests/api/test_account.py:639]: We should use `assert_not_called()` instead of call_count.

File: tests/api/test_account.py [hunk 4]
```diff
@@ -633,18 +636,24 @@ def test_customer_register(send_account_confirmation_email_mock, api_client):
...
 ):
-    variables = {"email": "customer@example.com", "password": "Password"}
+    email = "customer@example.com"
+    variables = {"email": email, "password": "Password"}
     response = api_client.post_graphql(ACCOUNT_REGISTER_MUTATION, variables)
     errors = response.json()["data"]["accountRegister"]["errors"]
+    new_user = User.objects.get(email=email)
+
     assert errors == []
     assert send_account_confirmation_email_mock.delay.call_count == 0
```
Comment (tomaszszymanski129) [inline tests/api/test_account.py:655]: Not your code but might be changed to assert_not_called()

File: saleor/order/utils.py [hunk 5]
```diff
@@ -333,3 +334,8 @@ def get_voucher_discount_for_order(order: Order) -> Money:
     if order.voucher.type == VoucherType.SPECIFIC_PRODUCT:
         return get_products_voucher_discount_for_order(order.voucher)
     raise NotImplementedError("Unknown discount type")
+
+
+def match_orders_with_new_user(user: User) -> None:
+    """Match Orders with a new user."""
```
Comment (tomaszszymanski129) [inline saleor/order/utils.py:340]: That line does explain nothing more to me than the func name :) maybe worth changing or reming entirely.

**Pull request merged**